File: kaisa/lib/ast.ml
module ParsedAst = struct
  type 'a node = {
    value: 'a;
    loc: int*int;
  }[@@deriving show]

  let mk_node value start_pos end_pos =
    {value; loc = (start_pos,end_pos)}

  type id = string
  [@@deriving show]
  and
    variable = id*id
    [@@deriving show]
  and
    value =
    | DefaultNumber of string node
    | DecimalNumber of string node
    | BinaryNumber of string node
    | HexNumber of string node
    | Character of string node
    | String of string node
    | List of expr list node
    | Tuple of expr list node
    [@@deriving show]
  and expr =
    | Id of string node
    | Val of value node
    | FuncCall of string*expr list
    | Operation of operation node
    | PatternMatch of expr*(expr list)*(expr) node
    [@@deriving show]
  and operation =
    | Binary of expr*binop*expr 
    | UnaryL of prefixop*expr 
    | UnaryR of expr*postfixop 
    [@@deriving show]
  and binop =
    | Plus
    | Minus
    | Times
    | Div
    | Access
  and prefixop =
    | Deref
    | Ref
    [@@deriving show]
  and postfixop =
    | Fact
    [@@deriving show]
  and stm =
    | LetVal of id
    | LetFn of id*(variable list option)*expr
    | LetType of id*(variable list option)*expr
    | LetIn of stm*(stm option)
  and ast = stm list
    [@@deriving show]
end

module SemanticAst = struct

end

File: src/frontend/ast.rs
use std::ops::Range;


#[derive(Debug,PartialEq, Eq)]
pub struct Node<T>
{
  loc: Range<usize>,
  literal: T,
}

impl<T> Node<T>
{
    pub fn new(loc: Range<usize>, lit: T) -> Self
    {
	Self
	{
	    loc,
	    literal: lit	
	}
    }
}

#[derive(Debug,PartialEq, Eq)]
pub enum Value<'a>
{
 Number(Node<&'a str>),
 Character(Node<&'a str>),
 String(Node<&'a str>),
 List(Node<Box<[Box<Expr<'a>>]>>),
 Tuple(Node<Box<[Box<Expr<'a>>]>>),
}

#[derive(Debug,PartialEq, Eq)]
pub enum Expr<'a>
{
 Id(Node<&'a str>),
 Val(Value<'a>),
 Binary(Box<Expr<'a>>,Binop,Box<Expr<'a>>),    
 Prefix(UnaryL,Box<Expr<'a>>),    
 Postfix(Box<Expr<'a>>,UnaryR),    
}

#[derive(Debug,PartialEq, Eq)]
pub enum Binop
{
 Plus(Node<()>),
 Minus(Node<()>),
 Times(Node<()>),
 Div(Node<()>),
 Access(Node<()>),
}

#[derive(Debug,PartialEq, Eq)]
pub enum UnaryL
{
 Ref(Node<()>), 
 Deref(Node<()>), 
 Minus(Node<()>), 
}

#[derive(Debug,PartialEq, Eq)]
pub enum UnaryR
{
 Excl(Node<()>), 
}
